torch/csrc/jit/passes/peephole_list_idioms.cpp

```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+// desugar to something similar. Additionally, any True refinements that were
+// present on `cond` can also be associated with the if node True output value.
+
+struct ListLenRefiner {
+  ListLenRefiner(
+      std::shared_ptr<Graph> graph,
+      std::unordered_set<Value*>& mutated_lists)
+      : graph_(std::move(graph)), mutated_lists_(mutated_lists) {}
+
+  bool run() {
+    std::unordered_set<Value*> li_with_len_use;
+    collectListsToRefine(graph_->block(), li_with_len_use);
+    if (lists_to_refine_.size() == 0) {
+      return false;
+    }
+    ListRefinement refinements;
+    RefineListLens(graph_->block(), refinements);
+    return changed_;
+  }
+
+  // we only need to analyze lists that have multiple uses of len(), and we can
+  // only analyze lists that are not mutated
+  void collectListsToRefine(
+      Block* b,
+      std::unordered_set<Value*>& li_with_len_use) {
+    for (Node* n : b->nodes()) {
+      for (Block* block : n->blocks()) {
+        collectListsToRefine(block, li_with_len_use);
+      }
+
+      if (n->kind() != aten::len) {
+        continue;
+      }
+
+      auto first_input = n->inputs().at(0);
+      if (first_input->type()->cast<ListType>() &&
+          !mutated_lists_.count(first_input)) {
+        if (!li_with_len_use.count(first_input)) {
+          li_with_len_use.insert(first_input);
+        } else {
+          lists_to_refine_.insert(first_input);
+        }
+      }
+    }
+  }
+
+  ListRefinement RefineListLens(Block* b, ListRefinement block_refinements) {
+    active_refinements_.push_back(&block_refinements);
+    for (Node* n : b->nodes()) {
+      if (n->matches("aten::eq(int a, int b) -> bool") ||
+          n->matches("aten::ne(int a, int b) -> bool")) {
+        // check for one input constant and the other coming from len(li)
+        for (size_t const_index : {0, 1}) {
+          auto ival = constant_as<int64_t>(n->inputs().at(const_index));
+          if (!ival) {
+            continue;
+          }
+          auto li_len = n->inputs().at(const_index - 1);
+          if (!li_len->node()->matches("aten::len.t(t[] a) -> int") ||
+              !lists_to_refine_.count(li_len->node()->input())) {
+            continue;
+          }
+          ListRefinement refine;
+          refine[li_len->node()->input()] = *ival;
+          info_[n->output()] = n->kind() == aten::eq
+              ? BoolRefinements::TrueRefinements(std::move(refine))
+              : BoolRefinements::FalseRefinements(std::move(refine));
+        }
+      }
+      if (n->kind() == prim::RaiseException) {
+        throwing_blocks_.insert(b);
+      }
+      if (n->kind() == aten::len) {
+        if (auto maybe_len = tryFindRefinement(n->inputs().at(0))) {
+          changed_ = true;
+          WithInsertPoint guard(n);
+          n->output()->replaceAllUsesWith(
+              graph_->insertConstant(static_cast<int64_t>(*maybe_len)));
+        }
+      }
+
+      if (n->kind() == prim::If) {
+        IfView if_n(n);
```

Comment:
Nit: might be worth factoring this out to a separate method.

Reply:
I'm doing this in a follow up PR to share logic between this and another pass so going to leave out of this one if that's fine.